### PR TITLE
Add banners view

### LINKED ISSUE #40: Implement Banner View
The banners that appear on the POS need to be editable for the CEB, on the /banners endpoint

### PR DESCRIPTION
**Pull request by SuperVK**
Add the bannerviews, includes but not limited to: Creating banners, updating banners, uploading new banner pictures, filtering on active/expired banners, viewing banners up close, mobile support.

### CODE REVIEW

**Comment by SuperVK** [comment]
TODO: Localisation

File: src/components/BannerDialogComponent.vue [hunk 1]
```diff
@@ -0,0 +1,216 @@
+<template>
+    <Dialog v-model:visible="visible" modal :draggable="false" :header="
+            banner != undefined
+            ? 'Edit banner'
```
Comment (CodeNamedRobin) [inline src/components/BannerDialogComponent.vue]: Raw text.

File: src/components/BannerDialogComponent.vue [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+            : 'Create banner'
```
Comment (CodeNamedRobin) [inline src/components/BannerDialogComponent.vue]: Raw text.

File: src/components/BannerDialogComponent.vue [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        </div>
+        <Divider />
+        <div class="py-2">
+            {{$t('banner.name')}} <br>
+            <InputText v-model="name" /> <br>
+            <span class="font-bold text-red-500">{{ errors['Name'] }}</span>
+        </div>
+        <div class="py-2">
+            {{ $t('banner.duration') }}<br>
+            <InputNumber v-model="duration" suffix=" seconds" :maxFractionDigits="3" /><br>
```
Comment (CodeNamedRobin) [inline src/components/BannerDialogComponent.vue]: Raw text.

File: src/components/BannerDialogComponent.vue [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    }
+});
+
+const visible = defineModel<boolean>('visible');
+// If Banner is undefined we are creating, otherwise editing.
+const banner = defineModel<BannerResponse | undefined>('banner');
+
+
+// Banner schema and validation
+// TODO: Actual validation and errors
```
Comment (CodeNamedRobin) [inline src/components/BannerDialogComponent.vue]: Make issue or fix.

File: src/components/BannerDialogComponent.vue [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+const bannerSchema = toTypedSchema(
+    yup.object({
+        'Name': yup.string().required(),
+        'Duration': yup.number().required().moreThan(0),
+        'Start date': yup.date().required(),
+        'End date': yup.date().required().min(new Date()).min(
+            yup.ref('Start date'),
+            "End date can't be before start date"
```
Comment (CodeNamedRobin) [inline src/components/BannerDialogComponent.vue:101]: Raw text.

File: src/components/BannerDialogComponent.vue [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+    }
+});
+
+const onImgUpload = (e: Event) => {
+    const el = (e.target as HTMLInputElement);
+    if (el == null || el.files == null) return;
+    uploadedImage.value = el.files[0]!!;
+};
+</script>
+<style>
```
Comment (CodeNamedRobin) [inline src/components/BannerDialogComponent.vue:182]: Why no inline style classes? :(

File: src/components/BannerItemComponent.vue [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        <div v-else class="px-3 py-5 text-xl surface-hover text-center">
+          {{ $t('banner.noBannersFound') }}
+        </div>
+        <!--
+                    Green when active and banner image is present
+                    Yellow when active but banner image is not present
+                    Red when not active
+                -->
+        <Tag
+          :value="banner.active ? 'Active' : 'Not active'"
```
Comment (CodeNamedRobin) [inline src/components/BannerItemComponent.vue]: Raw text.

File: src/components/BannerItemComponent.vue [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+          :severity="banner.active ? (banner.image ? 'success' : 'warning') : 'danger'"
+          class="absolute"
+          style="left: 4px; top: 4px"
+        />
+      </div>
+      <div class="flex flex-row justify-content-between w-full md:w-7">
+        <div class="flex flex-column pr-3">
+          <span class="text-xl">{{ banner.name }}</span
+          ><br />
+          <span class="font-italic">
```
Comment (CodeNamedRobin) [inline src/components/BannerItemComponent.vue]: These are not properly aligned around the horizontal axis. Could be fixed with some nice flexbox.

File: src/components/BannersListComponent.vue [hunk 9]
```diff
@@ -0,0 +1,68 @@
+<template>
+    <CardComponent header="Banner overview">
```
Comment (CodeNamedRobin) [inline src/components/BannersListComponent.vue]: Raw text

File: src/components/BannersListComponent.vue [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    banners: BannerResponse[],
+    skip?: number | undefined,
+    take?: number | undefined
+}>();
+
+const isCreateDialogVisible = ref<boolean>();
+
+const filters = ref<string[]>();
+
+const options = [
```
Comment (CodeNamedRobin) [inline src/components/BannersListComponent.vue:59]: Raw text

File: src/utils/formatterUtils.ts [hunk 11]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (CodeNamedRobin) [inline src/utils/formatterUtils.ts:1]: I smell a lint-fix

**Review by CodeNamedRobin — CHANGES_REQUESTED** [review]
Small issues, mostly raw text being used.

**Comment by CodeNamedRobin** [comment]
Also, the colors on switch button are quite hard to distinguish whether or not an option is chosen or not.

File: src/components/BannerDialogComponent.vue (hunk 5 above)
Comment (SuperVK) [reply src/components/BannerDialogComponent.vue:101]: #158

File: src/components/BannerDialogComponent.vue (hunk 6 above)
Comment (SuperVK) [reply src/components/BannerDialogComponent.vue:182]: This is more readable, also copied from primevue but adapted to work for my use case. Not the most pretty as editing products also uses this, but is an issue for later.

File: src/components/BannerItemComponent.vue (hunk 8 above)
Comment (SuperVK) [reply src/components/BannerItemComponent.vue]: Done!

File: src/utils/formatterUtils.ts (hunk 11 above)
Comment (SuperVK) [reply src/utils/formatterUtils.ts:1]: hm?

File: src/components/BannerDialogComponent.vue (hunk 5 above)
Comment (CodeNamedRobin) [reply src/components/BannerDialogComponent.vue:101]: Fair.

**Review by CodeNamedRobin — APPROVED** [review]
LGTM!

**Pull request merged**